service-base/src/test/java/org/eclipse/hono/service/credentials/AbstractCompleteCredentialsServiceTest.java

```diff
@@ -0,0 +1,257 @@
+/*******************************************************************************
+ * Copyright (c) 2016, 2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.hono.service.credentials;
+
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.unit.Async;
+import io.vertx.ext.unit.TestContext;
+import io.vertx.ext.unit.junit.VertxUnitRunner;
+import org.eclipse.hono.util.CredentialsConstants;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import java.net.HttpURLConnection;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
+
+/**
+ * Abstract class used as a base for verifying behavior of {@link CompleteCredentialsService} in device registry implementations.
+ *
+ */
+@RunWith(VertxUnitRunner.class)
```

Comment:
I would rather have the concrete test (sub-)classes contain this annotation in order to make it more explicit that it is really those (sub-)classes that implement the tests. WDYT?
Same applies to the timeout `@Rule` that is defined in one of the abstract base classes ...

Reply:
Make sense. Pushed the changes.